src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj

```diff
@@ -125,12 +150,12 @@
    shutdown-fn :- IFn]
   (future
     (let [stopped? (atom false)]
-      (while (not @stopped?)
-        (try
-          (let [[watch-key events] (retrieve-events watcher)]
-            (when-not (empty? events)
-              (process-events! watcher watch-key events shutdown-fn)))
-         (catch ClosedWatchServiceException e
-           (reset! stopped? true)
-           (log/info (trs "Closing watcher {0}" watcher))))))))
+      (shutdown-fn #(while (not @stopped?)
+                     (try
+                       (let [events (retrieve-events watcher)]
+                         (when-not (empty? events)
```

Comment:
The Oracle docs[1] say that `.pollEvents` could return an empty array. I think if that's the case then `get-events-from-watch-key` will return an empty seq, and if there's no additional events gathered in the loop that empty seq will be returned from `retrieve-events`.

I don't believe that that should happen unless we called `.pollEvents` on a `WatchKey` that hadn't been enqueued in the `WatchService`, but I haven't found docs that say definitively one way or the other. I should caveat that with I haven't looked too hard as that basic flow has existed from the initial implementation of this service.

1. [java.nio.file.WatchKey#pollEvents()](https://docs.oracle.com/javase/8/docs/api/java/nio/file/WatchKey.html#pollEvents--)